core/src/ldml/ldml_processor.cpp

```diff
@@ -231,75 +233,89 @@ ldml_processor::process_event(
       }
       break;
     default:
+    // all other VKs
       {
-        // adapted from kmx_processor.cpp
-
-        /** a copy of the current/changed context, for transform use */
-        ldml::string_list ctxt;
-
-        // Construct a context buffer of all the KM_KBP_BT_CHAR items
-        // Extract the context into 'ctxt' for transforms to process
-        if (!!transforms) {
-          // if no transforms, no reason to do this extraction (ctxt will remain empty)
-          auto &cp = state->context();
-          // We're only interested in as much of the context as is a KM_KBP_BT_CHAR.
-          uint8_t last_type = KM_KBP_BT_UNKNOWN;
-          for (auto c = cp.rbegin(); c != cp.rend(); c++) {
-            last_type = c->type;
-            if (last_type != KM_KBP_BT_CHAR) {
-              // not a char, stop here
-              // TODO-LDML: markers?
-              break;
-            }
-            ctxt.emplace_front(1, c->character);
-            // extract UTF-32 to 1 or 2 UTF-16 chars in a string
-          }
-        }
-
         // Look up the key
         const std::u16string str = keys.lookup(vk, modifier_state);
+
         if (str.empty()) {
-          // not found, so pass the keystroke on to the Engine
+          // no key was found, so pass the keystroke on to the Engine
           state->actions().push_invalidate_context();
           state->actions().push_emit_keystroke();
           break; // ----- commit and exit
         }
-        // found the correct string - push it into the context and actions
+
+        /**
+         * a copy of the current/changed context, for transform use.
+         *
+         */
+        std::u32string ctxtstr;
+
+        if (!!transforms) {
+          (void)context_to_string(state, ctxtstr);
+        } // else, don't bother populating
+
+        // found a string - push it into the context and actions
+        // we convert it here instead of using the emit_text() overload
+        // so that we don't have to reconvert it inside the transform code.
         const std::u32string str32 = kmx::u16string_to_u32string(str);
-        for (const auto &ch : str32) {
-          state->context().push_character(ch);
-          state->actions().push_character(ch);
-        }
-        // Now process transforms
-        // Process the transforms
+        emit_text(state, str32);
```

Comment:
By actually calling emit_text here it will cause the character to be added to the actions, this will bubble all the way up and the platform will output the character only for it to be then deleted in the case of a transform that requires deletion. It will work out ok but we will be sending the application a character then deleting it.
Instead, we can just let the transform have the new character, if there is no match at this point we add the new character. If there is a match and it wants to delete we don't have delete the first character as we had not yet added it. 

Also transform::apply states `@return match length: number of chars at end of input string to modify.  0 if no match.`
Is this correct that there are no transforms that simply result in two code points being emitted?
for example    `q -> qu`
If you can have this then in this case we will also want to add the new char and then apply the outputString

Reply:
the return from apply is 'the number of chars in the _matched_ text to change'

so if the input text was `pdq` and you apply `q -> qu` then the match length will be `1` but the output string `qu`, meaning to delete the `q` (size 1) and replace it with `qu` (size 2)

Reply:
Thanks for clarifying. It turns out the change I suggested and you made also works well then for this ` q -> qu` case as now it will just emit the `qu` without first adding and deleting it and then adding it again.